### PR TITLE
Flag for not overwrite js files by default without generating errors

### PR DESCRIPTION
**Pull request by zhengbli**
This PR adds a new internal flag to by default not overwrite js files, without generating the compiler diagnostic message. 

// cc @mhegazy @vladima

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -1537,13 +1537,19 @@ namespace ts {
                     const emitFilePath = toPath(emitFileName, currentDirectory, getCanonicalFileName);
                     // Report error if the output overwrites input file
                     if (filesByName.contains(emitFilePath)) {
-                        createEmitBlockingDiagnostics(emitFileName, Diagnostics.Cannot_write_file_0_because_it_would_overwrite_input_file);
+                        const sourceFile = filesByName.get(emitFilePath);
+                        if (isSourceFileJavaScript(sourceFile) && options.noEmitOverwriteForJsFiles) {
```
Comment (sheetalkamat) [inline src/compiler/program.ts]: Can you make this ```options.noEmitOverwriteForJsFiles && isSourceFileJavaScript(sourceFile)``` instead

Comment (vladima) [inline src/compiler/program.ts]: or maybe
```ts
blockEmittingOfFile(emitFileName, options.noEmitOverwriteForJsFiles && isSourceFileJavaScript(sourcFile) && Diagnostics.Cannot_write_file_0_because_it_would_overwrite_input_file);
```

File: src/compiler/types.ts [hunk 2]
```diff
@@ -2984,6 +2984,7 @@ namespace ts {
         moduleResolution?: ModuleResolutionKind;
         newLine?: NewLineKind;
         noEmit?: boolean;
+        /*@internal*/noEmitOverwriteForJsFiles?: boolean;
```
Comment (mhegazy) [inline src/compiler/types.ts]: nit. `noEmitOverwritenFiles`

File: src/compiler/program.ts [hunk 3]
```diff
@@ -1537,21 +1537,25 @@ namespace ts {
                     const emitFilePath = toPath(emitFileName, currentDirectory, getCanonicalFileName);
                     // Report error if the output overwrites input file
                     if (filesByName.contains(emitFilePath)) {
-                        let chain: DiagnosticMessageChain;
-                        if (!options.configFilePath) {
-                            // The program is from either an inferred project or an external project
-                            chain = chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Adding_a_tsconfig_json_file_will_help_organize_projects_that_contain_both_TypeScript_and_JavaScript_files_Learn_more_at_https_Colon_Slash_Slashaka_ms_Slashtsconfig);
+                        const sourceFile = filesByName.get(emitFilePath);
+                        if (options.noEmitOverwriteForJsFiles && isSourceFileJavaScript(sourceFile)) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: i would say we should do this for all files and not only JS files.

Comment (sheetalkamat) [inline src/compiler/program.ts]: What about --out = someJSSourceFile inferred? Should we not expect error here?

Comment (zhengbli) [inline src/compiler/program.ts]: actually that makes sense. That way you won't be surprised when opening a ts file along side a js file

File: src/compiler/program.ts [hunk 4]
```diff
@@ -1537,21 +1537,24 @@ namespace ts {
                     const emitFilePath = toPath(emitFileName, currentDirectory, getCanonicalFileName);
                     // Report error if the output overwrites input file
                     if (filesByName.contains(emitFilePath)) {
-                        let chain: DiagnosticMessageChain;
-                        if (!options.configFilePath) {
-                            // The program is from either an inferred project or an external project
-                            chain = chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Adding_a_tsconfig_json_file_will_help_organize_projects_that_contain_both_TypeScript_and_JavaScript_files_Learn_more_at_https_Colon_Slash_Slashaka_ms_Slashtsconfig);
+                        if (options.noEmitOverwritenFiles && !options.out && !options.outDir && !options.outFile) {
```
Comment (zhengbli) [inline src/compiler/program.ts]: That is true for now, though the relationship might be loose. It is possible that in the future this option can be enabled for other projects, while this checking guarantees correctness without costing much

**Comment by zhengbli** [comment]
@mhegazy @sheetalkamat @billti further comments?

**Comment by mhegazy** [comment]
@zhengbli can you port this to master as well.

**Pull request merged**